File: client/src/app/collector/linux/memory.py
from pathlib import Path


def _read_meminfo() -> dict[str, int]:
    result = {}
    for line in Path("/proc/meminfo").read_text().splitlines():
        key, _, value = line.partition(":")
        result[key.strip()] = int(value.strip().split()[0])
    return result


def memory_used_mb() -> float | None:
    try:
        info = _read_meminfo()
        total = info.get("MemTotal", 0)
        available = info.get("MemAvailable", 0)
        return round((total - available) / 1024, 1)
    except Exception:
        return None


def memory_total_mb() -> float | None:
    try:
        info = _read_meminfo()
        return round(info.get("MemTotal", 0) / 1024, 1)
    except Exception:
        return None

File: client/internal/collector/linux/memory.go
package linux

import (
	"context"
	"math"
	"os"
	"strconv"
	"strings"

	"statusphere-client/internal/collector"
	"statusphere-client/internal/presence"
)

func init() {
	collector.Register(collector.Descriptor{
		Provider: collector.Provider{Name: "memory", Collect: memory},
		Applies:  collector.OnOS("linux"),
	})
}

func parseMeminfo() (map[string]int64, error) {
	data, err := os.ReadFile("/proc/meminfo")
	if err != nil {
		return nil, err
	}

	result := make(map[string]int64)
	for _, line := range strings.Split(string(data), "\n") {
		key, val, ok := strings.Cut(line, ":")
		if !ok {
			continue
		}
		fields := strings.Fields(strings.TrimSpace(val))
		if len(fields) == 0 {
			continue
		}
		v, _ := strconv.ParseInt(fields[0], 10, 64)
		result[strings.TrimSpace(key)] = v
	}
	return result, nil
}

func memory(_ context.Context, snap presence.Snapshot) error {
	info, err := parseMeminfo()
	if err != nil {
		return err
	}
	total := info["MemTotal"]
	available := info["MemAvailable"]
	if total == 0 {
		return nil
	}
	snap.Set(presence.KeyMemUsedMB, math.Round(float64(total-available)/1024*10)/10)
	snap.Set(presence.KeyMemTotalMB, math.Round(float64(total)/1024*10)/10)
	return nil
}
